core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size()) {
+      channel.truncate(targetSize)
+      channel.position(targetSize)
```

Comment:
We do truncation in a couple of other places. (1) During recovery, the recovery thread will be the only writer. So, this is safe. (2) In the other case that we need to do truncation, we are switching a replica to a follower. In this case, we will first mark the replica as the follower and stop the existing replicaFetchThread if necessary. At this point, there won't be any write traffic to the log. Then, we will do do truncation. So, this is safe.